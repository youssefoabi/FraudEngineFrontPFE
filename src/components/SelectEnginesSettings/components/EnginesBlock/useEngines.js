import { filter, keys, map, pathOr } from 'ramda';
import { isNilOrEmpty } from 'ramda-adjunct';
import { useCallback, useMemo } from 'react';

import useEngines from '../../../../screens/EnginesManagement/useEngines';

const useEnginesList = ({ handleChange, enginesGroup }) => {
  const { engines: allEngines } = useEngines();
  const { engines: enginesGroupEngines, selectedEngine } = enginesGroup || {};

  const handleSelectEngine = ({ value }) => {
    const engine = pathOr({}, ['value'], value);
    const { name } = engine;
    handleChange({ id: 'selectedEngine', value: { ...engine, label: name } });
  };

  const addEngine = () => {
    if (isNilOrEmpty(selectedEngine)) return;

    const engineId = pathOr(null, ['id'], selectedEngine);

    handleChange({
      id: 'engines',
      value: {
        ...(enginesGroupEngines || {}),
        [engineId]: {
          engineId,
          engineName: pathOr('', ['name'], selectedEngine),
          order: keys(enginesGroupEngines).length,
        },
      },
    });
  };

  const mapEngine = useCallback(
    engine => {
      if (isNilOrEmpty(engine)) return null;

      const { name, id } = engine;
      const findEngine = pathOr(null, [id], enginesGroupEngines);
      const isNewEngine = findEngine === null;
      if (!isNewEngine) return null;

      return { value: { ...engine, engineId: id }, label: name };
    },
    [enginesGroupEngines],
  );

  const enginesSelect = useMemo(() => {
    const mappedEngines = map(mapEngine, allEngines);

    return filter(engine => engine !== null, mappedEngines);
  }, [allEngines, mapEngine]);

  return {
    engines: enginesSelect,
    handleSelectEngine,
    addEngine,
  };
};

export default useEnginesList;
