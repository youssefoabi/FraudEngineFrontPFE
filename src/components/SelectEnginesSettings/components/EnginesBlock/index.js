import { Button, DataEdit, GridItem, useStyles } from '@cdiscount/backoffice-ui';
import { func, shape } from 'prop-types';
import Add from '@material-ui/icons/Add';
import Grid from '@material-ui/core/Grid';
import React from 'react';

import EnginesList from './components/EnginesList';
import styles from './styles';
import useEngines from './useEngines';

const EnginesBlock = ({ enginesGroup, handleChange }) => {
  const { handleSelectEngine, engines: allEngines, addEngine } = useEngines({
    handleChange,
    enginesGroup,
  });
  const classes = useStyles(styles);

  return (
    <>
      <Grid container spacing={3}>
        <Grid item xs={6} md={4} lg={4}>
          <DataEdit
            fields={[
              {
                id: 'selectedEngine',
                type: 'select',
                label: 'Moteur',
                selectProps: {
                  options: allEngines,
                },
              },
            ]}
            values={enginesGroup}
            onChange={handleSelectEngine}
          />
        </Grid>
        <GridItem xs={6} md={3} lg={3}>
          <Button
            className={classes.ButtonWrapper}
            size="large"
            icon={Add}
            color="tertiary"
            variant="contained"
            label="AJOUTER UN ONGLET"
            onClick={addEngine}
          />
        </GridItem>
      </Grid>
      <EnginesList enginesGroup={enginesGroup} handleChange={handleChange} />
    </>
  );
};

EnginesBlock.propTypes = {
  enginesGroup: shape({}),
  handleChange: func,
};

EnginesBlock.defaultProps = {
  enginesGroup: null,
  handleChange: Function.Prototype,
};

export default EnginesBlock;
