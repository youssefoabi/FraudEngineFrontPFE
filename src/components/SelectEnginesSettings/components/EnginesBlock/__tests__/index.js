import { itRendersCorrectly } from '@cdiscount/jest-utils';
import React from 'react';

import EnginesBlock from '../';

jest.mock('react-beautiful-dnd', () => ({
  Droppable: ({ children }) =>
    children(
      {
        draggableProps: {
          style: {},
        },
        innerRef: jest.fn(),
      },
      {},
    ),
  Draggable: ({ children }) =>
    children(
      {
        draggableProps: {
          style: {},
        },
        innerRef: jest.fn(),
      },
      {},
    ),
  DragDropContext: ({ children }) => children,
}));

jest.mock('react-redux', () => ({
  useDispatch: () => ({}),
  useSelector: () => ({}),
}));

describe(EnginesBlock.name, () => {
  itRendersCorrectly(<EnginesBlock />);
});
