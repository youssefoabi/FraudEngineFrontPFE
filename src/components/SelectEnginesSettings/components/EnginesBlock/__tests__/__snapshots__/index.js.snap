// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EnginesBlock renders correctly 1`] = `
<Fragment>
  <WithStyles(ForwardRef(Grid))
    container={true}
    spacing={3}
  >
    <WithStyles(ForwardRef(Grid))
      item={true}
      lg={4}
      md={4}
      xs={6}
    >
      <f
        config={null}
        fields={
          Array [
            Object {
              "id": "selectedEngine",
              "label": "Moteur",
              "selectProps": Object {
                "options": Object {},
              },
              "type": "select",
            },
          ]
        }
        inputs={null}
        onChange={[Function]}
        onInit={[Function]}
        title={null}
        values={null}
      />
    </WithStyles(ForwardRef(Grid))>
    <h
      fullheight={false}
      lg={3}
      md={3}
      title={null}
      xs={6}
    >
      <y
        className="makeStyles-ButtonWrapper-1"
        color="tertiary"
        icon={
          Object {
            "$$typeof": Symbol(react.memo),
            "compare": null,
            "displayName": "AddIcon",
            "muiName": "SvgIcon",
            "type": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
          }
        }
        label="AJOUTER UN ONGLET"
        onClick={[Function]}
        size="large"
        variant="contained"
      />
    </h>
  </WithStyles(ForwardRef(Grid))>
  <EnginesList
    enginesGroup={null}
  />
</Fragment>
`;
